app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx

```diff
@@ -3,6 +3,7 @@ import {
   render,
   fireEvent,
   waitForElementToBeRemoved,
+  createEvent,
```

Comment:
@fdoxyz, on line 197,

```javascript
      expect(
        window.webkit.messageHandlers.imageUpload.postMessage,
      ).toHaveBeenCalledTimes(1);
```

it's failing because `window.ForemMobile` does not exist. Was this working before as this code appears to be unchanged.

Instead of testing the internals of `window.ForemMobile.injectNativeMessage(namespace, message)` opt to test `window.ForemMobile.injectNativeMessage` being called as this is the public API.

Reply:
This failure is also the cause of another failure (false negative). If this is sorted, that should be sorted.

```
<ArticleCoverImage /> › when rendered in native iOS with imageUpload support › when an image is uploaded › successfully uploads an image

    TypeError: _this.props.onMainImageUrlChange is not a function

      120 |         break;
      121 |       case 'success':
    > 122 |         this.props.onMainImageUrlChange({ links: [message.link] });
          |                    ^
      123 |         this.setState({ uploadingImage: false });
      124 |         break;
      125 |     }

      at Document.ArticleCoverImage._this.handleNativeMessage (app/javascript/article-form/components/ArticleCoverImage.jsx:122:20)
      at node_modules/@testing-library/preact/node_modules/@testing-library/dom/dist/events.js:25:20
      at err (node_modules/@testing-library/preact/dist/pure.js:55:16)
      at Object.<anonymous>.exports.act (node_modules/preact/test-utils/src/index.js:96:9)
      at Object.eventWrapper (node_modules/@testing-library/preact/dist/pure.js:54:30)
      at fireEvent (node_modules/@testing-library/preact/node_modules/@testing-library/dom/dist/events.js:16:35)
      at Object.<anonymous> (app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx:224:9)
      at asyncGeneratorStep (app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx:17:103)
      at _next (app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx:19:194)
      at app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx:19:364
      at Object.<anonymous> (app/javascript/article-form/components/__tests__/ArticleCoverImage.test.jsx:19:97)
```

Reply:
I don't think `base.jsx` gets executed inside these jest tests. Since that isn't initialized then the namespaced functions aren't available. I just substituted the mocked globals with `jest.fn()` to mock/spy the ForemMobile namespace functions instead of the webkit handlers directly.

> Was this working before as this code appears to be unchanged.

Yes because we weren't doing dynamic importing. The real problem is that the dynamic import happens in `base.jsx` and it isn't executed here. I think the mocks in that regard fixed those broken tests along with the `JSON.stringify` update 🙌 

> TypeError: _this.props.onMainImageUrlChange is not a function

It seems like this is the only one left, I don't think it sorted the other one up. Not sure why this one breaks if we're passing in a `jest.fn()` as a `onMainImageUrlChange` prop. I'm going to keep looking into this a bit more and try to figure it out.

Reply:
This is all sorted in [`5988ce2` (#15212)](https://github.com/forem/forem/pull/15212/commits/5988ce2fa3c74bdadebef5a3df1cc02fa05a6b88)